File: myAtoi.py
"""
TC O(N) ---> O(len of maxint) 
SC O(1)

"""

class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.strip()

        if not s or ((not s[0].isdigit()) and s[0] != '+' and s[0] != '-'):
            return 0

      
        sign = '-' if s[0] == '-' else '+'

        # Set limits to prevent overflow
        max_limit = 214748364 #integer.MAX_VALUE to prevent overflow  2**31 - 1
        res = 0
        
        
        """
        Positive Number :
        If current result is > maxint/10, any incoming digit will cause overflow - return maxint
        If current result == maxint/10, the incoming digit must be < 7, if it is > 7, - return maxint
        
        Negative Number:
        If current result is > max_int/10, any incoming digit will cause overflow - return -max_int
        If current result == max_int/10, the incoming digit must be < 8, if it is > 8, - return -max_int
        
        
        """

       
        for i in range(len(s)):
            #get char at idx i
            c = s[i]
            if c.isdigit():#if digit then covert to str
                int_c = ord(c) - ord('0')
                if sign == '+':  # positive number
                    if res > max_limit or (res == max_limit and int_c > 7):
                        return 2147483647 #maxlimit 2**31-1
                else:  # negative number
                 
                    if res > max_limit or (res == max_limit and int_c > 8):
                        return -2147483648 #min limimt - 2**31
                # No overflow, add curr digit to result
                res = res * 10 + int_c
                """
                Handle non digit characters within the string.Only the first ch can be a non digit - '+' or '-'. If any other char is a non digit, break, res will contain the int formed by the string till this point in code.
                
                """
          
            elif i != 0:
                break

        # Check sign and return result
        return res if sign == '+' else -res
        

File: MyAtoi.java
// Time Complexity : The time complexity is O(n) where n is the length of the string
// Space Complexity : Te space complexity is O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class Solution {
    public int myAtoi(String s) {

        s = s.trim();

        if(s == null || s.length() == 0) return 0;

        int result = 0;
        char ch = s.charAt(0);
        //if the first character is a letter
        if(ch != '+' && ch != '-' && !Character.isDigit(ch)) return 0;

        char sign = '+';

        if(ch == '-'){
            sign = '-';
        }

        int limit = Integer.MAX_VALUE/10;

        for(int i=0;i<s.length();i++){

            ch = s.charAt(i);

            if(Character.isDigit(ch)){

                //if the number if positive
                if(sign == '+'){
                    //if the number goes out of limits
                    if(result > limit) return Integer.MAX_VALUE;
                    else if(result == limit && (ch - '0') >= 7) return Integer.MAX_VALUE;
                }
                //if the number is negative
                else if(sign == '-'){
                    //if the number goes out of limits
                    if(result > limit) return Integer.MIN_VALUE;
                    else if(result == limit && (ch - '0') >= 8) return Integer.MIN_VALUE;
                }
                result = result*10 + ch - '0';
            }
            else if(i != 0){
                break;
            }
        }

        if(sign == '+') return result;
        return -result;

    }
}